README.md

```diff
@@ -2,6 +2,324 @@
 
 Dependency injection for Glimmer applications.
 
+## What is Dependency Injection?
+
+Dependency injection is a pattern that increases the flexibility, testability
+and consistency of your code.
+
+The three key ideas are:
+
+1. An object's dependencies should be provided to the object when it is created,
+   rather than hardcoded.
+2. A dependency may have multiple implementations, so long as each
+   implementation adheres to an agreed-upon interface.
+3. An object using a dependency shouldn't care where on the filesystem that
+   dependency comes from.
+
+When we talk about an object's dependencies, we just mean the other objects that
+it relies on to do its job.
+
+Let's look at a short example that **does not** use dependency injection. We'll
+write a hypothetical server that renders a short HTML document when an incoming
+request is received:
+
+```js
+import HTTPServer from "./servers/http";
+
+export default class HelloWorldServer {
+  constructor() {
+    let server = new HTTPServer({
+      port: 80
+    });
+
+    server.on('request', req => {
+      req.write("<html><body>Hello, world!</body></html>");
+    });
+  }
+}
+```
+
+This is great, but there's one problem. As you can see, our Hello World server
+is importing the HTTP library directly. If we want to support both HTTP and
+HTTP/2 (or even something like a WebSocket), this code is not reusable.
+
+We would have to either duplicate this code, or add some configuration options
+to let the user tell us which protocol they want to use. Of course, that would
+work today, but if we wanted to support HTTP/3 in the future, we'd have to come
+back and add a new configuration option for every new protocol.
+
+What if, instead of telling the server what protocol to use, we could instead
+provide it with an object that encapsulated all of those concerns? Instead of
+having our `HelloWorldServer` import an `HTTPServer` directly, we can provide it
```

Comment:
This reads correctly to my ear (@dgeb?) but I'm open to a clearer way to express this.